File: oa_questions/arrays/163MissingRange.js
/**
 * @param {number[]} nums
 * @param {number} lower
 * @param {number} upper
 * @return {string[]}
 */
 var findMissingRanges = function(nums, lower, upper) {
    const addRange = (lower, upper) => {
        return `${lower}->${upper}`
    }
    const addNum = n => String(n)
    let output = []

    nums = [lower - 1, ...nums, upper + 1]

    for(let i = 0; i<nums.length-1; i++) {
        const lowerBound = nums[i] + 1
        const upperBound = nums[i+1] -1
        if(nums[i+1] - nums[i] !== 1) {
            if(lowerBound === upperBound) output.push(addNum(lowerBound))
            else output.push(addRange(lowerBound, upperBound))
        }
    }
    return output
};

File: oa_questions/arrays/163MissingRange.py
class 163MissingRange:
    def findMissingRanges(self, nums: List[int], lower: int, upper: int) -> List[str]:
        output = []

        nums = [lower - 1] + nums + [upper + 1]

        for i in range(len(nums) -1):
            lowerBound = nums[i] + 1
            upperBound = nums[i+1] -1
            if nums[i+1] - nums[i] != 1:
                if lowerBound == upperBound:
                    output.append(f"{lowerBound}")
                else:
                    output.append(f"{lowerBound}->{upperBound}")
        return output

class 163MissingRange:
    def findMissingRanges(self, nums: List[int], lower: int, upper: int) -> List[str]:
        output = []

        nums = nums + [upper + 1]

        previous = lower - 1

        for num in nums:
            if num == previous + 2:
                output.append(f"{previous+1}")
            elif num > previous + 2:
                output.append(f"{previous+1}->{num-1}")

            previous = num

        return output